### PR TITLE
[PLT-0] LBOX Namespace, Plugin Style SDK

### PR DESCRIPTION
**Pull request by adrian-chang**
# Description

https://labelbox.atlassian.net/wiki/spaces/PLT/pages/2658074642/TDD+Labelbox+Python+Repo+Organization

Labelbox lbox example + initial develop.yaml additions

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Document change (fix typo or modifying any markdown files, code comments or anything in the examples folder only)

## All Submissions

- [x] Have you followed the guidelines in our Contributing document?
- [x] Have you provided a description?
- [x] Are your changes properly formatted?

## New Feature Submissions

- [x] Does your submission pass tests?
- [x] Have you added thorough tests for your new feature?
- [x] Have you commented your code, particularly in hard-to-understand areas?
- [x] Have you added a Docstring?

## Changes to Core Features

- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully run tests with your changes locally?
- [x] Have you updated any code comments, as applicable?

### CODE REVIEW

**Comment by vbrodsky** [comment]
I've reviewed the wonderful wiki documentation [[Wiki Link]](https://labelbox.atlassian.net/wiki/spaces/PLT/pages/2658074642/TDD+Labelbox+Python+Repo+Organization) and skimmed through the pull request related to the proposed changes for the SDK.  Based on my understanding it seems the proposal aims to address several objectives:

- Enable safe contributions of external packages to the SDK by external developers.
- Facilitate workload distribution between different subteams within the SDK development process.
- Potentially introduce a mechanism for deprecating SDK features.

**Complexity Concerns**:  While I appreciate the potential benefits of these changes, I'm concerned that the proposed solution might lead to an increase in overall SDK development complexity in the near future.  Personally, I would prefer to minimize such complexity.

Before moving forward with this specific proposal, I believe it would be valuable to take a step back and identify the most pressing challenges we need to address for the SDK's future development.  By focusing on the core problems, we can potentially explore alternative solutions that achieve our goals without introducing unnecessary complexity.

I'd be happy to discuss this further and collaborate on defining a clear roadmap for the SDK's future.

**Comment by adrian-chang** [comment]
> Complexity Concerns: While I appreciate the potential benefits of these changes, I'm concerned that the proposed solution might lead to an increase in overall SDK development complexity in the near future. Personally, I would prefer to minimize such complexity.

I'd like hear this concern about complexity. I think its much simpler to operate in a bunch of micro / smaller packages (although there is acknowledged organic overhead of just doing more than a single package) than to deal with a single monolithic SDK package. You get versioning, the ability to do small breaking changes, run a small subset of tests, and the flexibility to do a lot more things than having to worry about changes to code in other areas of the codebase.

Breaking things up is always a tough thing to do, but in fact we really aren't overtly breaking this up for the sake of breaking things up. In fact, I wouldn't really even consider this breaking up the SDK, more about enforcing strong boundaries been modules where that doesn't exist at all currently. 

To be 100% clear, there **are no required or mandatory near future changes.** SDK3 can be developed as SDK3 (this PR makes no impact on it).  Any of these changes can be back-ported to SDK3 also.

Happy to discuss, but the main objective is to organically move things over in time to assist to getting an SDK4.

> Before moving forward with this specific proposal, I believe it would be valuable to take a step back and identify the most pressing challenges we need to address for the SDK's future development.

If you feel there is a pressing need somewhere else, please create a proposal and execute on it and we can compare then.

File: .github/actions/lbox-matrix/action.yml [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  files-changed:
+    description: 'Files changed as json array'
+    required: true
+    default: '[]'
+outputs:
+  matrix: 
+    description: 'Matrix to run lbox packages tests against'
+runs:
+  using: 'node20'
+  main: 'index.js'
```
Comment (vbrodsky) [inline .github/actions/lbox-matrix/action.yml:13]: what does this file do

File: .github/workflows/lbox-develop.yml [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        run: rye format --check -v -p ${{ matrix.package }}
+      - name: Linting
+        run: rye lint -v -p ${{ matrix.package }}
+      - name: Unit
+        working-directory: libs/${{ matrix.package }}
+        run: rye run unit
+      - name: Integration
+        working-directory: libs/${{ matrix.package }}
+        env:
+          LABELBOX_TEST_API_KEY: ${{ secrets[matrix.api-key] }}
```
Comment (vbrodsky) [inline .github/workflows/lbox-develop.yml:69]: where are these matrix attribures (api-key, package, etc) defined?

File: .github/workflows/lbox-develop.yml [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+          VERSION=$(date +"%Y.%m.%d.%H.%M")
+          echo "pip install --index-url https://test.pypi.org/simple/ --extra-index-url=https://pypi.org/simple/ ${{ matrix.package }}@$VERSION" >> "$GITHUB_STEP_SUMMARY"
+          rye version "$VERSION"
+          rye run toml set --toml-path pyproject.toml project.name  ${{ matrix.package }}
+          rye build
+      - name: Publish package distributions to Test PyPI
+        uses: pypa/gh-action-pypi-publish@release/v1
+        with:
+          packages-dir: dist/
+          repository-url: https://test.pypi.org/legacy/
```
Comment (vbrodsky) [inline .github/workflows/lbox-develop.yml:108]: why is repo url /legacy?

**Review by vbrodsky — COMMENTED** [review]
approvable, just had a couple of questions

File: .github/workflows/lbox-develop.yml (hunk 3 above)
Comment (adrian-chang) [reply .github/workflows/lbox-develop.yml:108]: Open the URL in the browser, it's the pypi api dns upload endpoint.

Comment (adrian-chang) [reply .github/workflows/lbox-develop.yml:108]: In this case, for test-pypi

File: .github/workflows/lbox-develop.yml (hunk 2 above)
Comment (adrian-chang) [reply .github/workflows/lbox-develop.yml:69]: See .github/actions/lbox-matrix/index.js

  const startingMatrix = [
    {
      "python-version": "3.8",
      "api-key": "STAGING_LABELBOX_API_KEY_2",
      "da-test-key": "DA_GCP_LABELBOX_API_KEY"
    },
    {
      "python-version": "3.9",
      "api-key": "STAGING_LABELBOX_API_KEY_3",
      "da-test-key": "DA_GCP_LABELBOX_API_KEY"
    },
    {
      "python-version": "3.10",
      "api-key": "STAGING_LABELBOX_API_KEY_4",
      "da-test-key": "DA_GCP_LABELBOX_API_KEY"
    },
    {
      "python-version": "3.11",
      "api-key": "STAGING_LABELBOX_API_KEY",
      "da-test-key": "DA_GCP_LABELBOX_API_KEY"
    },
    {
      "python-version": "3.12",
      "api-key": "STAGING_LABELBOX_API_KEY_5",
      "da-test-key": "DA_GCP_LABELBOX_API_KEY"
    }
  ];

File: .github/actions/lbox-matrix/action.yml (hunk 1 above)
Comment (adrian-chang) [reply .github/actions/lbox-matrix/action.yml:13]: It's a github custom github action definition https://docs.github.com/en/actions

Comment (vbrodsky) [reply .github/actions/lbox-matrix/action.yml:13]: does it have to be a 26K+ js file?

**Review by vbrodsky — APPROVED** [review]

**Pull request merged**